core/jvmMain/src/kotlinx/serialization/internal/Platform.kt

```diff
@@ -153,6 +153,9 @@ private fun <T : Any> Class<T>.createEnumSerializer(): KSerializer<T> {
 }
 
 private fun <T : Any> Class<T>.findObjectSerializer(): KSerializer<T>? {
+    // Special case to avoid IllegalAccessException on Java11+ (#2449)
+    // There are no serializable objects in the stdlib anyway.
+    if (this.canonicalName.let { it.startsWith("java.") || it.startsWith("kotlin.") }) return null
```

Comment:
The problem is that it is not accessible (`IllegalAccessException`), if it is not accessible then it is also not an object. See the suggested additions to the test suite (JavaCollectionsTest) that are not fixed by the solution.
```suggestion
    if (!Modifier.isPublic(this.modifiers)) return null
```

Reply:
This check will not work correctly for `protected` situation like this one:
```
open class A {
  @Serializable protected object B {}
}
```

Regarding `@Serializable private object X` — you're right; this code will throw an IllegalAccessError now. However, it will indicate problems with users' code instead of masking it behind the `null`. Also, we support `@Serializable private class X` even if it is in a different package. So, probably, it should be supported instead of returning null. Seems like it is out of the scope of this PR anyways.

Reply:
Maybe the "best" way would be to just catch the `IllegalAccessException` as it is otherwise really hard to know what is accessible.

Reply:
It's an adequate approach (although we lose information that is contained in expectation message). However, I should first check if we can support `@Serializable private object X` properly.